### PR TITLE
feat(google iap): add iap support for user login

### PR DESCRIPTION
**Pull request by ybai2sc**
add IAP support for auth, add support for Spinnaker instances which has IAP enabled for auth

### CODE REVIEW

File: cmd/gateclient/client.go [hunk 1]
```diff
@@ -31,19 +31,20 @@ import (
 	"path/filepath"
 	"strings"
 
+	"github.com/spinnaker/spin/config"
+	config2 "github.com/spinnaker/spin/config/auth/iap"
```
Comment (jtk54) [inline cmd/gateclient/client.go]: Is there a particular reason this is aliased? Shouldn't this resolve to the `iap` package? If the alias is necessary I'd name it `iap` or something similar.

File: config/auth/iap/oauth.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+func (o *oauthReceiver) killWhenReady(n net.Conn, s http.ConnState) {
+	if o.killMeNow && s.String() == "idle" && o.mainClient == n.RemoteAddr().String() {
+		o.doneChan <- nil
+	}
+}
+
+// NewOAuthConfig creates a new OAuth config
+func (o *oauthReceiver) NewOAuthConfig() *oauth2.Config {
+
+	// return oauth2 Config
```
Comment (jtk54) [inline config/auth/iap/oauth.go]: I think this is self-documenting.

File: config/auth/iap/oauth.go [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+	oauthCode := r.FormValue("code")
+	state := r.FormValue("state")
+	if oauthCode == "" || state == "" {
+		response := fmt.Errorf("Invalid response from Google's Account Service")
+		o.WebOutput(w, "Uh Oh!", response.Error())
+		o.doneChan <- response
+		return
+	}
+
+	var validateStateFunc func(string, string) bool
```
Comment (jtk54) [inline config/auth/iap/oauth.go]: Necessary to save this in a local variable?

Comment (ybai2sc) [reply config/auth/iap/oauth.go]: thanks, have removed this local variable

File: config/auth/iap/oauth.go (hunk 2 above)
Comment (ybai2sc) [reply config/auth/iap/oauth.go]: thx, removed this comment

File: cmd/gateclient/client.go (hunk 1 above)
Comment (ybai2sc) [reply cmd/gateclient/client.go]: changed the package name to `iap`

**Pull request merged**